File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
        int maxArea = 0;
        int l = 0;
        int r = heights.length - 1;
        while (l < r) {
            maxArea = Math.max(maxArea, (r - l) * Math.min(heights[r], heights[l]));
            if (heights[l] < heights[r]) {
                l += 1;
            } else if (heights[l] >= heights[r]) {
                r -= 1;
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-2.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # Return the minimum eating rate k
        # So basically we go from 1 -> onward
        # to find the hours where we could eat the banana within h (<= h)

        # Brute force solution: creates res = 1
        # iterate through the piles, at each index, digest everything at res ,
        # keep track of hours
        # until the end of the array, see if hours <= h, if not, keep going

        # The upper bound for k is max(piles). Because if Koko eats the largest
        # pile in 1 go (hour), she could do the same for the rest
        if not piles:
            return -1
  
        l, r = 1, max(piles)

        while l <= r:
            m = (r + l) // 2 # Let m be the current speed to test

            totalTime = 0
            for pile in piles:
                totalTime += math.ceil(float(pile) / m)
            
            if totalTime <= h:
                res = m
                # Keep finding smaller possible time, as we're looking for the min
                r = m - 1
            else:
                l = m + 1
        
        return res